File: dynamicProgramming/minPathSum.js
function minPathSum(grid) {
  var height = grid.length;
  var width = grid[0].length;

  for (var i = 0; i < height; i++) {
    for (var j = 0; j < width; j++) {
      if (i !== 0 && j !== 0)
        grid[i][j] += Math.min(grid[i - 1][j], grid[i][j - 1]);
      else if (i !== 0) grid[i][j] += grid[i - 1][j];
      else if (j !== 0) grid[i][j] += grid[i][j - 1];
    }
  }
  return grid[height - 1][width - 1];
}

File: dynamicProgramming/minPathSum.py
def minPathSum(grid):
    h, w = len(grid), len(grid[0])
    dp_table = grid

    for x in range(1, w):
        dp_table[0][x] = grid[0][x] + dp_table[0][x-1]

    for y in range(1, h):
        dp_table[y][0] = grid[y][0] + dp_table[y-1][0]

    for y in range(1, h):
        for x in range(1, w):
            dp_table[y][x] = grid[y][x] + \
                min(dp_table[y][x-1], dp_table[y-1][x])

    return dp_table[h-1][w-1]

# https://leetcode.com/problems/minimum-path-sum/submissions/
